core/shared/src/main/scala/zio/ZIO.scala

```diff
@@ -1043,21 +1043,38 @@ sealed trait ZIO[-R, +E, +A]
   final def once(implicit trace: Trace): UIO[ZIO[R, E, Unit]] =
     Ref.make(true).map(ref => self.whenZIO(ref.getAndSet(false)).unit)
 
+  /**
+   * Executes the specified success or error callback depending on the result of
+   * this effect, synchronously completing the effect before returning.
+   */
   final def onDone[R1 <: R](
     error: E => ZIO[R1, Nothing, Any],
     success: A => ZIO[R1, Nothing, Any]
   )(implicit trace: Trace): ZIO[R1, Nothing, Unit] =
-    ZIO.uninterruptibleMask { restore =>
-      restore(self).foldZIO(e => restore(error(e)), s => restore(success(s))).forkDaemon.unit
-    }
+    onExit {
+      case Exit.Success(value) => success(value).unit
+      case Exit.Failure(cause) =>
+        cause.failureOrCause
+          .fold(
+            error,
+            _ => ZIO.unit
+          )
+          .unit
+    }.catchAll(_ => ZIO.unit).as(())
```

Comment:
Reuse `onDoneCause`:

```suggestion
    onDoneCause(
      _.failureOrCause.fold(error, Exit.failCause(_)),
      success
    )
```